### PR TITLE
node/bindnode: more minor improvements for graphsync

### PR DESCRIPTION
**Pull request by mvdan**
(see commit messages - please do not squash)

### CODE REVIEW

**Comment by mvdan** [comment]
Note to self: revisit bindnode's Next key being valid only until the next call.

File: node/basicnode/string.go [hunk 1]
```diff
@@ -13,8 +13,12 @@ var (
 )
 
 func NewString(value string) datamodel.Node {
-	v := plainString(value)
-	return &v
+	return plainString(value)
+}
+
+// NewStringAddr is like NewString, but it does not allocate a new string.
```
Comment (warpfork) [inline node/basicnode/string.go:19]: A short comment here about how this lets you make a mutable node, and warning that that might not be what you or others expect, would make me happier.

File: node/bindnode/repr.go [hunk 2]
```diff
@@ -1026,7 +1026,8 @@ func (w *_structIteratorRepr) Next() (key, value datamodel.Node, _ error) {
 		keyStr, _ := key.AsString()
 		mappedKey := outboundMappedKey(stg, keyStr)
 		if mappedKey != keyStr {
-			key = basicnode.NewString(mappedKey)
+			w.keyStringAddr = mappedKey
+			key = basicnode.NewStringAddr(&w.keyStringAddr)
```
Comment (warpfork) [inline node/bindnode/repr.go:1030]: Similarly, a short comment on this iterator (or somewhere that's discoverable?) about how it's *not safe* to use this node elsewhere and assume that it's immutable seems very important.  The norm is to assume that a node is immutable and you can stitch it together in larger COW structures without second thoughts.

**Review by warpfork — COMMENTED** [review]
Most of these make sense.

I concede on the `TypeStruct.Fields` accessor and `TypeEnum.Members` accessor for sure.  (I keep wanting Golang to have immutablity.  It doesn't.  I should stop wishing I can have nice things; and paying extra for them isn't worth it.)

I'm fairly nervous about the iterators returning something as a Node that's mutable.  Maybe with enough documentation, it's going to be survivable.  Still nerve-wracking and a fairly large divergence from the usual contract.

I wish we could make the iterators for structs and unions (e.g. what's wishing for these string wrappers) just grab pointers to "constant" strings that the `schema.Type` values already contain.  That's hard at present because _that_ package also spent effort on trying to be immutable in a language that's inimically hostile to it... but maybe it's worth drilling there instead.

File: node/bindnode/infer.go [hunk 3]
```diff
@@ -285,7 +285,13 @@ func inferGoType(typ schema.Type) reflect.Type {
 }
 
 // from IPLD Schema field names like "foo" to Go field names like "Foo".
-func fieldNameFromSchema(name string) string {
+func inferFieldNameFromSchema(name string) string {
+	// Hard-code a few common names to be inferred better.
+	// TODO: is there a good and small library collecting such names?
+	switch name {
+	case "id":
+		return "ID"
```
Comment (rvagg) [inline node/bindnode/infer.go:293]: I'm _meh_ about this one, feels like a stretch to be doing this. `Id` isn't the worst thing to end up with is it? If you go down this path you're looking at having an entry for every known acronym out there.

Comment (warpfork) [reply node/bindnode/infer.go:293]: +1 tbh.  I don't relish anything that gives me flashbacks to Ruby/Rail's "Inflector".

Comment (mvdan) [reply node/bindnode/infer.go:293]: I definitely don't want to maintain this long-term, but without it, graphsync's schema doesn't quite work. Unless we rename its Go `ID` to `Id`, which feels unnecessary.

We can definitely do better on our end, and that's https://github.com/ipld/go-ipld-prime/issues/329. I don't want to block us for another two weeks, hence the temporary internal hack.

Comment (mvdan) [reply node/bindnode/infer.go:293]: On second thought, it sounds like on graphsync I'll be using separate types for bindnode, so I don't need to worry about `Id` vs `ID` anymore. I'll undo this.

Comment (mvdan) [reply node/bindnode/infer.go]: the commit is now entirely removed.

**Comment by mvdan** [comment]
After talking to @warpfork about it yesterday, I ended up backing out my `NewStringAddr` change; it breaks the `Next` semantics from the MapIterator interface, and also makes us return `ipld.Node`s which get modified under the hood, which is likely going to cause bugs.

All that doesn't seem worth it for a minor speed-up right now. I might investigate other avenues which don't lose us safety in the near future.

**Comment by mvdan** [comment]
> I concede on the `TypeStruct.Fields` accessor and `TypeEnum.Members` accessor for sure. (I keep wanting Golang to have immutablity. It doesn't. I should stop wishing I can have nice things; and paying extra for them isn't worth it.)

Taking this to mean that you're OK with that one commit, and given that the other two were removed, I'm going ahead and merging this :)

**Pull request merged**